SmartTutor/ContentModel/LearningObjects/Challenges/FulfillmentStrategy/ProjectChecker/ProjectAnalyzer.cs

```diff
@@ -0,0 +1,70 @@
+﻿using CodeModel.CaDETModel.CodeItems;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SmartTutor.ContentModel.LearningObjects.Challenges.FulfillmentStrategy.ProjectChecker
+{
+    public class ProjectAnalyzer : ChallengeFulfillmentStrategy
+    {
+        public Dictionary<string, List<ChallengeFulfillmentStrategy>> SnippetApplicableStrategies { get; private set; }
+
+        private ProjectAnalyzer() { }
+
+        public ProjectAnalyzer(Dictionary<string, List<ChallengeFulfillmentStrategy>> snippetApplicableStrategies) : this()
+        {
+            SnippetApplicableStrategies = snippetApplicableStrategies;
+        }
+
+        public override HintDirectory EvaluateSubmission(List<CaDETClass> solutionAttempt)
+        {
+            var challengeHints = new HintDirectory();
+            foreach (var strategy in SnippetApplicableStrategies.SelectMany(sas => sas.Value))
+            {
+                var result = strategy.EvaluateSubmission(GetAllClassesFromSnippets(solutionAttempt));
+                challengeHints.MergeHints(result);
+            }
+            return challengeHints;
+        }
+
+        public override List<ChallengeHint> GetAllHints()
+        {
+            return SnippetApplicableStrategies.SelectMany(sas => sas.Value).SelectMany(cfs => cfs.GetAllHints().Where(h => h != null)).ToList();
```

Comment:
We can remove the Where(h => h != null) as the Challenge already performs this check.